sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/executeImmediate.scala

```diff
@@ -0,0 +1,199 @@
+      }
+    }
+  }
+
+  def extractQueryString(either: Either[String, UnresolvedAttribute]): String = {
+    either match {
+      case Left(v) => v
+      case Right(u) =>
+        val varReference = getVariableReference(u.nameParts)
+
+        if (!varReference.dataType.sameType(StringType)) {
+          throw new AnalysisException(
+            errorClass = "INVALID_VARIABLE_TYPE_FOR_QUERY_EXECUTE_IMMEDIATE",
+            messageParameters = Map("varType" -> varReference.dataType.simpleString))
+        }
+
+        // Call eval with null value passed instead of a row.
+        // This is ok as this is variable and invoking eval should
+        // be independent of row value.
+        varReference.eval(null).toString
+    }
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan =
+    plan.resolveOperatorsWithPruning(_.containsPattern(EXECUTE_IMMEDIATE), ruleId) {
+      case ExecuteImmediateQuery(expressions, query, targetVariablesOpt) =>
+        val queryString = extractQueryString(query)
+        val plan =
+          parseStatement(catalogManager.v1SessionCatalog.parser, queryString, targetVariablesOpt)
+
+        val posNodes = plan.collect { case p: LogicalPlan =>
+          p.expressions.flatMap(_.collect { case n: PosParameter => n })
+        }.flatten
+        val namedNodes = plan.collect { case p: LogicalPlan =>
+          p.expressions.flatMap(_.collect { case n: NamedParameter => n })
+        }.flatten
+
+        val queryPlan = if (expressions.isEmpty || (posNodes.isEmpty && namedNodes.isEmpty)) {
+          plan
+        } else if (posNodes.nonEmpty && namedNodes.nonEmpty) {
+          throw new AnalysisException(
+            errorClass = "INVALID_QUERY_BOTH_POSITIONAL_AND_NAMED_PARAMETERS_PRESENT",
+            messageParameters = Map.empty)
+        } else {
+          if (posNodes.nonEmpty) {
+            PosParameterizedQuery(
+              plan,
+              // We need to resolve arguments before Resolution batch to make sure
+              // that some rule does not accidently resolve our parameters.
+              // We do not want this as they can resolve some unsupported parameters
+              resolveArguments(expressions))
+          } else {
+            val namedExpressions = expressions.collect { case (e: NamedExpression) => e }
+
+            NameParameterizedQuery(
+              plan,
+              namedExpressions.map(_.name),
+              // We need to resolve arguments before Resolution batch to make sure
+              // that some rule does not accidently resolve our parameters.
+              // We do not want this as they can resolve some unsupported parameters.
+              resolveArguments(namedExpressions))
+          }
+        }
+
+        targetVariablesOpt
+          .map(variables => {
+            SetVariable(variables, queryPlan)
+          })
+          .getOrElse { queryPlan }
+    }
+
+  private def parseStatement(
+      parser: ParserInterface,
```

Comment:
nit: we don't need to pass it. We can get `catalogManager.v1SessionCatalog.parser` within this method